File: merge_sort/merge_sort.py
from __future__ import print_function

def mergeSort(array, start, end):
    if(start < end):
        n = (end - start) + 1
        middle = int((end + start)/2)
        mergeSort(array, start, middle)
        mergeSort(array, middle + 1, end)
        merge(array, start, end, middle, n)

def merge(array, start, end, middle, n):
    array2 = [ None ] * n
    i = start
    j = middle + 1
    for k in range(start, n):
        if(i > middle):
            array2[k] = array[j]
            j += 1
        elif(j > end):
            array2[k] = array[i]
            i += 1
        elif(array[i] < array[j]):
            array2[k] = array[i]
            i += 1
        else:
            array2[k] = array[j]
            j += 1

    for k in range(start, n):
        array[k] = array2[k]

def main():
    array = [4, 8, 9, 15, 1, 2, 6]
    mergeSort(array, 0, len(array) - 1)
    for i in array:
        print(i, end=' ')
    print()

main()

File: merge_sort/merge_sort.js


function mergeSort(array){
   var len = array.length;
   if(len < 2){
     return array;
   }
    
   var middle, left, right;
   middle = Math.floor(len/2)
   left = array.slice(0, middle);
   right = array.slice(middle);
   
   return merge(mergeSort(left),mergeSort(right));
}
        


function merge(left, right){
  var result = [],
      lLen = left.length,
      rLen = right.length;

  var leftPtr = 0, rightPtr = 0;
      
  while(leftPtr < lLen && rightPtr < rLen){
     if(left[leftPtr] < right[rightPtr]){
       result.push(left[leftPtr]);
       leftPtr += 1;
     }
     else{
       result.push(right[rightPtr]);
       rightPtr += 1;
    }
  }  

  return result.concat(left.slice(leftPtr)).concat(right.slice(rightPtr));

}

t = [1,2,4,5,6,2,3,8,12,43]
x = [2,4,5]
console.log(mergeSort(t))
console.log(mergeSort(x))
